Update Sorting_Custom.mdx (#1012)
cpinitiative/usaco-guide: content/3_Silver/Sorting_Custom.mdx

So, with prefix sums and coordinate compression, it is possible for us to find the value of each special interval. Now, let's finish the problem. Let the "sum of a special interval" be the sum of every element inside the special interval, or in other words, the interval's length times its value. We already know each special interval's value, but we can also get its length because we stored the original indices. The length of an interval is its right index minus its left index, which we can find by accessing the sorted list. As a result, we can also find the sum of each special interval with the information we have. When the problem queries us for the sum between two special indices, we add up the sums of the special intervals between those two special indices. Thus, if we create another prefix sum array on the sums of the special intervals, we can easily answer these range sum queries in O(1) each.

The code below can be expanded to show the full solution.

</Spoiler>

<LanguageSection>

<CPPSection>

```cpp
//BeginCodeSnip{Solution code}
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;

> ll difference_array[400005]; //difference_array[i] = the difference of the values between special intervals i-1 and i

Reviewer: ```suggestion
ll difference_array[400005]; 
//difference_array[i] = the difference of the values between special intervals i-1 and i
```

		widths[i+1] = indices[i+1] - indices[i];
	}

	//We use prefix sums of the difference array to get the values of the intervals
	for(int i = 1; i < indices.size(); i++) {
		interval_value[i] = interval_value[i-1] + difference_array[i];
	}

	//We use prefix sums over the sums of the special intervals to be able to answer queries quickly
	for(int i = 1; i < indices.size(); i++) {
		prefix_sums[i] = prefix_sums[i-1] + interval_value[i] * widths[i];
	}

	//Classic use of prefix sum array to answer range sum queries
	for(int i = 0; i < Q; i++) {
		cout << prefix_sums[getCompressedIndex(queries[i].second)] - prefix_sums[getCompressedIndex(queries[i].first)] << "\n";
	}

	return 0;
}
//EndCodeSnip
```

</CPPSection>

<JavaSection>

```java
//BeginCodeSnip{Solution code}
import java.util.*;
import java.io.*;

public class rangequeries {
	
	//custom class that stores a few integers (used for directly storing the queries and updates given in the input)
	static class Query {
		
		int l, r, v;
		public Query(int l, int r, int v) {
			this.l = l;
			this.r = r;
			this.v = v;
		}
		
		public Query(int l, int r) {
			this.l = l;
			this.r = r;
		}
		
	}
	

> 	static long difference_array[]; //difference_array[i] = the difference of the values between special intervals i-1 and i

Reviewer: ```suggestion
	static long difference_array[]; 
	//difference_array[i] = the difference of the values between special intervals i-1 and i
```

	static int getCompressedIndex(int a) {
		return Collections.binarySearch(indices, a);
	}
//BeginCodeSnip{Solution code}
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); // A Scanner is too slow to solve this problem!
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken()), Q = Integer.parseInt(st.nextToken());
		
		indices = new ArrayList<Integer>();
		queries = new Query[Q];
		updates = new Query[N];
		
		for(int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			int l = Integer.parseInt(st.nextToken()), r = Integer.parseInt(st.nextToken()), v = Integer.parseInt(st.nextToken());
			indices.add(l);
			indices.add(r);
			updates[i] = new Query(l, r, v);
		}
		for(int i = 0; i < Q; i++) {
			st = new StringTokenizer(br.readLine());
			int l = Integer.parseInt(st.nextToken()), r = Integer.parseInt(st.nextToken());
			indices.add(l);
			indices.add(r);
			queries[i] = new Query(l, r);
		}
//EndCodeSnip
		//========= COORDINATE COMPRESSION =======

> 		TreeSet<Integer> temp = new TreeSet<Integer>(indices); //Since temp is a set, all duplicate elements are removed

Reviewer: ```suggestion
		TreeSet<Integer> temp = new TreeSet<Integer>(indices); 
		//Since temp is a set, all duplicate elements are removed
```

		indices.clear(); 
		indices.addAll(temp); //The elements of temp will be added to indices in sorted order, since temp is a TreeSet

> 		//The effect of these lines is to sort indices while removing duplicate elements. While this is not the only way to do it, it is one that requires little code.

Reviewer: ```suggestion
		//The effect of these lines is to sort indices while removing duplicate elements. 
		//While this is not the only way to do it, it is one that requires little code.
```

</LanguageSection>

## Coordinate Compression

Coordinate compression describes the process of mapping each value in a list to its index if that
list was sorted. For example, the list $\{7, 3, 4, 1\}$ would be compressed to
$\{3, 1, 2, 0\}$. Notice that $1$ is the least value in the first list, so it
becomes $0$, and $7$ is the greatest value, so it becomes $3$, the largest index
in the list.

When we have values from a large range, but we
only care about their relative order (for example, if we have to know if one

> value is above another), coordinate compression is a simple way to help a lot

Reviewer 2: ```suggestion
value is above another), coordinate compression is a simple way to help
```

with implementation. For example, if we have a set of integers ranging from $0$ to $10^9$, we

> can't use them as array indices because we'd have to create an array of size

Reviewer 2: ```suggestion
won't be able to use them as array indices because we'd have to create an array of size
```

$10^9$, which would surely cause a `Memory Limit Exceeded` verdict. However, if

> there are only $N \leq 10^6$ such integers, we can coordinate-compress their

Reviewer 2: ```suggestion
there are only $N \leq 10^6$ such integers, we can use coordinate compression to compress
```

Reviewer 3: i think the former is better

Reviewer: ^

values, which guarantees that the values will all be in the range from $0$ to

> $N-1$, which _can_ be used as array indices.

Reviewer 2: ```suggestion
$N-1$, which won't result in `Memory Limit Exceeded`.
```

<FocusProblem problem="sample2" />

> A good example of coordinate compression in action is in the solution of _USACO

Reviewer 4: ```suggestion
### Example 1

A good example of coordinate compression in action is in the solution of _USACO
```

		for (int x = 0; x <= n; x++) {
			for (int y = 0; y <= n; y++) {
				if (x > 0) {
					sums[x][y] += sums[x - 1][y];
				}
				if (y > 0) {
					sums[x][y] += sums[x][y - 1];
				}
				if (x > 0 && y > 0) {
					sums[x][y] -= sums[x - 1][y - 1];
				}
			}
		}
		long answer = n + 1;
		for (int j = 0; j < n; j++) {
			for (int k = j + 1; k < n; k++) {
				answer += getSum(Math.min(xs[j], xs[k]), Math.max(xs[j], xs[k]), 1, Math.min(ys[j], ys[k]))
					* getSum(Math.min(xs[j], xs[k]), Math.max(xs[j], xs[k]), Math.max(ys[j], ys[k]), n);
			}
		}
		System.out.println(answer);
	}
}
//EndCodeSnip
```

The solution uses a [lambda](https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html) function as the custom comparator, which our guide
didn't discuss, but it should be apparent which coordinate (`x` or `y`) that the
comparator is sorting by.

</JavaSection>

</LanguageSection>

The solution to Rectangular Pasture directly replaces coordinates with their compressed values, and forgets the real values of the coordinates because they are unnecessary. However, there may be problems for which we need to also remember the original values of coordinates that we compress.

> <FocusProblem problem="sample3" />

Reviewer 4: ```suggestion
### Example 2

<FocusProblem problem="sample3" />
```

value is above another), coordinate compression is a simple way to help a lot
with implementation. If we have a set of integers ranging from $0$ to $10^9$, we
can't use them as array indices because we'd have to create an array of size
$10^9$, which would surely cause a `Memory Limit Exceeded` verdict. However, if
there are only $N \leq 10^6$ such integers, we can coordinate-compress their
values, which guarantees that the values will all be in the range from $0$ to
$N-1$, which _can_ be used as array indices.

<FocusProblem problem="sample2" />

A good example of coordinate compression in action is in the solution of _USACO
Rectangular Pasture._ Again, we won't delve into the full solution but instead
discuss how coordinate compression is applied. Since the solution uses
[2D-prefix](/silver/prefix-sums-2) sums (another Silver topic), it is helpful if
all point coordinates are coordinate-compressed to the range $0$ to $N-1$ so
they can be used as array indices.

Below you will find the solution code to Rectangular Pasture, which uses
coordinate compression at the beginning. Observe how a custom comparator is used to sort the
points:

<LanguageSection>

<CPPSection>

```cpp

> BeginCodeSnip{}
> #include <iostream>
> #include <algorithm>
> using namespace std;
>  
> typedef pair<int,int> Point;
> bool ycomp(Point p, Point q) { return p.second < q.second; }
>  
> const int MAX_N = 2500;
> int N, Psum[MAX_N+1][MAX_N+1];
> Point P[MAX_N];
>  
> int rsum(int x1, int y1, int x2, int y2)
> {
>   return Psum[x2+1][y2+1] - Psum[x2+1][y1] - Psum[x1][y2+1] + Psum[x1][y1];
> }
>  
> int main(void)
> {
>   cin >> N;
>   for (int i=0; i<N; i++) {
>     int x, y;
>     cin >> x >> y;
>     P[i] = make_pair(x,y);
>   }
> EndCodeSnip
>   sort(P, P+N); 
>   for (int i=0; i<N; i++) P[i].first = i+1;
>   sort(P, P+N, ycomp); 
>   for (int i=0; i<N; i++) P[i].second = i+1;
> BeginCodeSnip{}
>   for (int i=0; i<N; i++) Psum[P[i].first][P[i].second] = 1;
>   for (int i=1; i<=N; i++)
>     for (int j=1; j<=N; j++)
>       Psum[i][j] += Psum[i-1][j] + Psum[i][j-1] - Psum[i-1][j-1];
>   long long answer = 0;
>   for (int i=0; i<N; i++) 
>     for (int j=i; j<N; j++) {
>       int x1 = min(P[i].first, P[j].first) - 1;
>       int x2 = max(P[i].first, P[j].first) - 1;
>       answer += rsum(0,i,x1,j) * rsum(x2,i,N-1,j);
>     }
>   cout << answer + 1 << "\n";
>   
> }
> EndCodeSnip
> 

Reviewer 4: ```suggestion
//BeginCodeSnip{Solution code}
#include <iostream>
#include <algorithm>
using namespace std;
//EndCodeSnip
 
typedef pair<int,int> Point;
bool ycomp(Point p, Point q) { return p.second < q.second; }

//BeginCodeSnip{Solution code}
const int MAX_N = 2500;
int N, Psum[MAX_N+1][MAX_N+1];
Point P[MAX_N];
 
int rsum(int x1, int y1, int x2, int y2)
{
	return Psum[x2+1][y2+1] - Psum[x2+1][y1] - Psum[x1][y2+1] + Psum[x1][y1];
}
 
int main(void)
{
	cin >> N;
	for (int i=0; i<N; i++) {
		int x, y;
		cin >> x >> y;
		P[i] = make_pair(x,y);
	}
//EndCodeSnip

	sort(P, P+N); 
	for (int i=0; i<N; i++) P[i].first = i+1;
	sort(P, P+N, ycomp); 
	for (int i=0; i<N; i++) P[i].second = i+1;

//BeginCodeSnip{Solution code}
	for (int i=0; i<N; i++) Psum[P[i].first][P[i].second] = 1;
	for (int i=1; i<=N; i++)
		for (int j=1; j<=N; j++)
			Psum[i][j] += Psum[i-1][j] + Psum[i][j-1] - Psum[i-1][j-1];
	long long answer = 0;
	for (int i=0; i<N; i++) 
		for (int j=i; j<N; j++) {
			int x1 = min(P[i].first, P[j].first) - 1;
			int x2 = max(P[i].first, P[j].first) - 1;
			answer += rsum(0,i,x1,j) * rsum(x2,i,N-1,j);
		}
	cout << answer + 1 << "\n";
	
}
//EndCodeSnip
```

Reviewer 4: Maybe this looks better?

Author: AGreed

```

</CPPSection>

<JavaSection>

```java
BeginCodeSnip{}
import java.util.Arrays;

> import java.util.Comparator;
> import java.util.Scanner;
>  
> public class RectangularPasture {
>     static int[][] sums;
>  
>     static int getSum(int fromX, int toX, int fromY, int toY) {
>         return sums[toX][toY] - sums[fromX - 1][toY] - sums[toX][fromY - 1] + sums[fromX - 1][fromY - 1];
>     }
>  
>     public static void main(String[] args) {
>         Scanner in = new Scanner(System.in);
>         int n = in.nextInt();
>         int[] xs = new int[n];
>         int[] ys = new int[n];
>         Integer[] cows = new Integer[n];
>         for (int j = 0; j < n; j++) {
>             xs[j] = in.nextInt();
>             ys[j] = in.nextInt();
>             cows[j] = j;
>         }
> EndCodeSnip
>         Arrays.sort(cows, Comparator.comparingInt(j -> xs[j]));
>         for (int x = 1; x <= n; x++) {
>             xs[cows[x - 1]] = x;
>         }
>         Arrays.sort(cows, Comparator.comparingInt(j -> ys[j]));
>         for (int y = 1; y <= n; y++) {
>             ys[cows[y - 1]] = y;
>         }
> BeginCodeSnip{}
>         sums = new int[n + 1][n + 1];
>         for (int j = 0; j < n; j++) {
>             sums[xs[j]][ys[j]]++;
>         }
>         for (int x = 0; x <= n; x++) {
>             for (int y = 0; y <= n; y++) {
>                 if (x > 0) {
>                     sums[x][y] += sums[x - 1][y];
>                 }
>                 if (y > 0) {
>                     sums[x][y] += sums[x][y - 1];
>                 }
>                 if (x > 0 && y > 0) {
>                     sums[x][y] -= sums[x - 1][y - 1];
>                 }
>             }
>         }
>         long answer = n + 1;
>         for (int j = 0; j < n; j++) {
>             for (int k = j + 1; k < n; k++) {
>                 answer += getSum(Math.min(xs[j], xs[k]), Math.max(xs[j], xs[k]), 1, Math.min(ys[j], ys[k]))
>                         * getSum(Math.min(xs[j], xs[k]), Math.max(xs[j], xs[k]), Math.max(ys[j], ys[k]), n);
>             }
>         }
>         System.out.println(answer);
>     }
> }
> EndCodeSnip
> 

Reviewer 4: ```suggestion
//BeginCodeSnip{Solution code}
import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

public class RectangularPasture {
	static int[][] sums;

	static int getSum(int fromX, int toX, int fromY, int toY) {
		return sums[toX][toY] - sums[fromX - 1][toY] - sums[toX][fromY - 1] + sums[fromX - 1][fromY - 1];
	}

	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int n = in.nextInt();
		//EndCodeSnip

		int[] xs = new int[n];
		int[] ys = new int[n];

		Integer[] cows = new Integer[n];
		for (int j = 0; j < n; j++) {
			xs[j] = in.nextInt();
			ys[j] = in.nextInt();
			cows[j] = j;
		}

		Arrays.sort(cows, Comparator.comparingInt(j -> xs[j]));
		for (int x = 1; x <= n; x++) {
			xs[cows[x - 1]] = x;
		}
		Arrays.sort(cows, Comparator.comparingInt(j -> ys[j]));
		for (int y = 1; y <= n; y++) {
			ys[cows[y - 1]] = y;
		}

//BeginCodeSnip{Solution code}
		sums = new int[n + 1][n + 1];
		for (int j = 0; j < n; j++) {
			sums[xs[j]][ys[j]]++;
		}
		for (int x = 0; x <= n; x++) {
			for (int y = 0; y <= n; y++) {
				if (x > 0) {
					sums[x][y] += sums[x - 1][y];
				}
				if (y > 0) {
					sums[x][y] += sums[x][y - 1];
				}
				if (x > 0 && y > 0) {
					sums[x][y] -= sums[x - 1][y - 1];
				}
			}
		}
		long answer = n + 1;
		for (int j = 0; j < n; j++) {
			for (int k = j + 1; k < n; k++) {
				answer += getSum(Math.min(xs[j], xs[k]), Math.max(xs[j], xs[k]), 1, Math.min(ys[j], ys[k]))
					* getSum(Math.min(xs[j], xs[k]), Math.max(xs[j], xs[k]), Math.max(ys[j], ys[k]), n);
			}
		}
		System.out.println(answer);
	}
}
//EndCodeSnip
```

I think this looks better

$\{3, 1, 2, 0\}$. Notice that $1$ is the least value in the first list, so it
becomes $0$, and $7$ is the greatest value, so it becomes $3$, the largest index
in the list.

When we have values that can take on a large range of possible values, but we
only care about their relative order (for example, if we have to know if one
value is above another), coordinate compression is a simple way to help a lot
with implementation. If we have a set of integers ranging from $0$ to $10^9$, we
can't use them as array indices because we'd have to create an array of size
$10^9$, which would surely cause a `Memory Limit Exceeded` verdict. However, if
there are only $N \leq 10^6$ such integers, we can coordinate-compress their
values, which guarantees that the values will all be in the range from $0$ to
$N-1$, which _can_ be used as array indices.

<FocusProblem problem="sample2" />

A good example of coordinate compression in action is in the solution of _USACO
Rectangular Pasture._ Again, we won't delve into the full solution but instead
discuss how coordinate compression is applied. Since the solution uses
[2D-prefix](/silver/prefix-sums-2) sums (another Silver topic), it is helpful if

> all point coordinates are coordinate-compressed to the range $0$ to $N-1$ so

Reviewer: add smth about otherwise overflow or memory exceed for array declaration and tle running 2d psums on that or similar??

they can be used as array indices.

> Below you will find the solution code to Rectangular Pasture, which uses

Reviewer: ```suggestion
Below you will find the solution to Rectangular Pasture, which uses
```

coordinate compression at the beginning. Observe how a custom comparator is used to sort the
points:

<LanguageSection>

<CPPSection>

```cpp
BeginCodeSnip{}

> #include <iostream>

Reviewer 2: Totally up to you, but `#include <bits/stdc++.h>` would be better...

Reviewer 4: Fine as is

#include <algorithm>
using namespace std;
 
typedef pair<int,int> Point;
bool ycomp(Point p, Point q) { return p.second < q.second; }
 
const int MAX_N = 2500;
int N, Psum[MAX_N+1][MAX_N+1];
Point P[MAX_N];
 
int rsum(int x1, int y1, int x2, int y2)
{
  return Psum[x2+1][y2+1] - Psum[x2+1][y1] - Psum[x1][y2+1] + Psum[x1][y1];
}
 
int main(void)
{
  cin >> N;
  for (int i=0; i<N; i++) {
    int x, y;
    cin >> x >> y;
    P[i] = make_pair(x,y);
  }
EndCodeSnip
  sort(P, P+N); 
  for (int i=0; i<N; i++) P[i].first = i+1;

>   sort(P, P+N, ycomp); 

Reviewer 2: Also up to you: `sort(P, P+N, [](Point a, Point b){
           return a.second < b.second;
}
);`

Reviewer 4: Fine as is.

		int n = in.nextInt();
		//EndCodeSnip

		int[] xs = new int[n];
		int[] ys = new int[n];

		Integer[] cows = new Integer[n];
		for (int j = 0; j < n; j++) {
			xs[j] = in.nextInt();
			ys[j] = in.nextInt();
			cows[j] = j;
		}

		Arrays.sort(cows, Comparator.comparingInt(j -> xs[j]));
		for (int x = 1; x <= n; x++) {
			xs[cows[x - 1]] = x;
		}
		Arrays.sort(cows, Comparator.comparingInt(j -> ys[j]));
		for (int y = 1; y <= n; y++) {
			ys[cows[y - 1]] = y;
		}

//BeginCodeSnip{Solution code}
		sums = new int[n + 1][n + 1];
		for (int j = 0; j < n; j++) {
			sums[xs[j]][ys[j]]++;
		}
		for (int x = 0; x <= n; x++) {
			for (int y = 0; y <= n; y++) {
				if (x > 0) {
					sums[x][y] += sums[x - 1][y];
				}
				if (y > 0) {
					sums[x][y] += sums[x][y - 1];
				}
				if (x > 0 && y > 0) {
					sums[x][y] -= sums[x - 1][y - 1];
				}
			}
		}
		long answer = n + 1;
		for (int j = 0; j < n; j++) {
			for (int k = j + 1; k < n; k++) {
				answer += getSum(Math.min(xs[j], xs[k]), Math.max(xs[j], xs[k]), 1, Math.min(ys[j], ys[k]))
					* getSum(Math.min(xs[j], xs[k]), Math.max(xs[j], xs[k]), Math.max(ys[j], ys[k]), n);
			}
		}
		System.out.println(answer);
	}
}
//EndCodeSnip
```

> The solution uses a lambda function as the custom comparator, which our guide

Reviewer: could prob link a lambda orcale docs

Reviewer: maybe this? https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html

    }
  cout << answer + 1 << "\n";
  
}

EndCodeSnip

```

</CPPSection>

<JavaSection>

```java

BeginCodeSnip{}
import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;
 
public class RectangularPasture {
    static int[][] sums;
 
    static int getSum(int fromX, int toX, int fromY, int toY) {
        return sums[toX][toY] - sums[fromX - 1][toY] - sums[toX][fromY - 1] + sums[fromX - 1][fromY - 1];
    }
 
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[] xs = new int[n];
        int[] ys = new int[n];
        Integer[] cows = new Integer[n];
        for (int j = 0; j < n; j++) {
            xs[j] = in.nextInt();
            ys[j] = in.nextInt();
            cows[j] = j;
        }
        // ------------- (coordinate compression) ----------
        Arrays.sort(cows, Comparator.comparingInt(j -> xs[j]));
        for (int x = 1; x <= n; x++) {
            xs[cows[x - 1]] = x;
        }
        Arrays.sort(cows, Comparator.comparingInt(j -> ys[j]));
        for (int y = 1; y <= n; y++) {
            ys[cows[y - 1]] = y;
        }

> 	// -------------------------------------------------

Reviewer 4: What i meant is that you can snip out everything except this portion.

Reviewer 4: The point is you are displaying the "important" part and hiding the rest. If someone wants to see the full implementation, they can expand the snips. Otherwise, only the part we want to emphasize should be shown.

Author: I see, lemme try that

    }
}
EndCodeSnip

```

The solution uses a lambda function as the custom comparator, which our guide
didn't discuss, but it should be apparent which coordinate (`x` or `y`) that the
comparator is sorting by.

</JavaSection>

</LanguageSection>

The problem statement of Rectangular Pasture ensures that all given `x` and `y` coordinates are distinct. However, coordinate compression can also be applied to problems where values are repeated and where we want to remove duplicate values. The following code shows ways of setting up coordinate compression with the removal of duplicate values:

<LanguageSection>

<CPPSection>

```cpp
BeginCodeSnip{}
cmp = v; //"cmp" starts as a copy of a vector v, but will end up with the elements of v in sorted order with duplicates removed
sort(begin(cmp), end(cmp)); //v is a vector of a type for which no custom comparator is necessary, though you can add a custom comparator to this sort() call as necessary
cmp.erase(unique(begin(cmp), end(cmp)), end(cmp)); 
auto cid = [&](int x) { return lower_bound(begin(cmp), end(cmp), x) - begin(cmp); }; // cid gives the index of x in cmp by "binary searching"
EndCodeSnip
```

</CPPSection>

<JavaSection>

```java

> BeginCodeSnip{}

Reviewer 4: Like it would be unnecessary here -- there's nothing "useless" in a sense

</LanguageSection>

## Coordinate Compression

> Coordinate compression is another useful technique related to sorting that can

Reviewer 5: You can delete the first sentence to make it more concise.

also be used in conjuction with custom comparators. Coordinate compression
describes the process of replacing each value in a list with its index if that
list was sorted. For example, the list $\{7, 3, 4, 1\}$ would be compressed to
$\{3, 1, 2, 0\}$. Notice that $1$ is the least value in the first list, so it
becomes $0$, and $7$ is the greatest value, so it becomes $3$, the largest index
in the list.

When we have values that can take on a large range of possible values, but we
only care about their relative order (for example, if we have to know if one
value is above another), coordinate compression is a simple way to help a lot
with implementation. If we have a set of integers ranging from $0$ to $10^9$, we
can't use them as array indices because we'd have to create an array of size
$10^9$, which would surely cause a `Memory Limit Exceeded` verdict. However, if

> there are only $N = 10^6$ such integers, we can coordinate-compress their

Reviewer 5: N < 10^6 would be better imo

values, which guarantees that the values will all be in the range from $0$ to
$N-1$, which _can_ be used as array indices.

<FocusProblem problem="sample2" />

A good example of coordinate compression in action is in the solution of _USACO
Rectangular Pasture._ Again, we won't delve into the full solution but instead
discuss how coordinate compression is applied. Since the solution uses

> [2D-prefix](/silver/prefix-sums-2) sums (another Silver topic), it is helpful if

Reviewer 5: the only reason I don't particularly like this example is bc it doesn't handle cases where there doesn't have to be unique coordinates. Although this problem covers that, I'm sure there are others that don't. Not sure if the other reviewers have an opinion on this.

</LanguageSection>

> 

Reviewer: what about the coordinate compression where spacing should be maintained asw well? tho i'm not sure if those showed up in silver yet.

When we have values that can take on a large range of possible values, but we only care about their relative order (for example, if we have to know if one value is above another), coordinate compression is a simple way to help a lot with implementation. If we have a set of integers ranging from $0$ to $10^9$, we can't use them as array indices because we'd have to create an array of size $10^9$, which would surely cause a `Memory Limit Exceeded` verdict. However, if there are only $N = 10^6$ such integers, we can coordinate-compress their values, which guarantees that the values will all be in the range from $0$ to $N-1$, which *can* be used as array indices.

<FocusProblem problem={problems.sample[1]} />

> A good example of coordinate compression in action is in the solution of *USACO Rectangular Pasture.* Again, we won't delve into the full solution but instead discuss how coordinate compression is applied. Since the solution uses 2D-prefix sums (another Silver topic), it is helpful if all point coordinates are coordinate-compressed to the range $0$ to $N-1$ so they can be used as array indices.

Reviewer 2: Maybe try linking the 2D Prefix Sums module.

  }
  sort(P, P+N); 
  for (int i=0; i<N; i++) P[i].first = i+1;
  sort(P, P+N, ycomp); 
  for (int i=0; i<N; i++) P[i].second = i+1;

...

```

</CPPSection>

<JavaSection>

```java

import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;
 
public class RectangularPasture {

...

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[] xs = new int[n];
        int[] ys = new int[n];
        Integer[] cows = new Integer[n];
        for (int j = 0; j < n; j++) {
            xs[j] = in.nextInt();
            ys[j] = in.nextInt();
            cows[j] = j;
        }
        Arrays.sort(cows, Comparator.comparingInt(j -> xs[j]));
        for (int x = 1; x <= n; x++) {
            xs[cows[x - 1]] = x;
        }
        Arrays.sort(cows, Comparator.comparingInt(j -> ys[j]));
        for (int y = 1; y <= n; y++) {
            ys[cows[y - 1]] = y;
        }

...

```

The solution uses a lambda function as the custom comparator, which our guide didn't discuss, but it should be apparent which coordinate (`x` or `y`) that the comparator is sorting by.

</JavaSection>

> <PySection>

Reviewer: i mean deleting the py section ad a whole will auto do an incomplete section

When we have values that can take on a large range of possible values, but we only care about their relative order (for example, if we have to know if one value is above another's), coordinate compression is a simple way to help a lot with implementation. If we have a set of integers ranging from $$0$$ to $$10^9$$, we can't use them as array indices because we'd have to create an array of size $$10^9$$, which would surely cause a `Memory Limit Exceeded` verdict. However, if there are only $$N = 10^6$$ such integers, we can coordinate compress their values, which guarantees that the values will all be in the range from $$0$$ to $$N-1$$, which *can* be used as array indices.

<FocusProblem problem={problems.sample[1]} />

A good example of coordinate compression in action is in the solution of *USACO Rectangular Pasture.* Again, we won't delve into the full solution but instead discuss how coordinate compression is applied. Since the solution uses 2D-prefix sums (another Silver topic), it is helpful if all point coordinates are coordinate-compressed to the range $$0$$ to $$N-1$$ so they can be used as array indices.

> Below you will find an excerpt of solution code to Rectangular Pasture that uses coordinate compression. Observe how a custom comparator is used to sort the points (either by `x` or by `y`):

Reviewer 4: Single dollar sign for inline latex

<LanguageSection>

<CPPSection>

```cpp

#include <iostream>
#include <algorithm>
using namespace std;

typedef pair<int,int> Point;
bool ycomp(Point p, Point q) { return p.second < q.second; }

const int MAX_N = 2500;
Point P[MAX_N];

> ...

Reviewer 4: I think it'd be quite nice to have the full soln but snipped out via code snips.

Author: Sorry, could you describe what you mean by code snips?

Reviewer 3: @nchn27 https://github.com/cpinitiative/usaco-guide/blob/master/docs/Content%20Documentation.md#code-blocks-and-code-snippets

Reviewer 4: About those docs... I don't think it properly mentions that commenting out is unnecessary -- it just searches for any instance of "BeginCodeSnip" or "CodeSnip" in each line (the first, if there are multiple for some reason)

import java.util.Comparator;
import java.util.Scanner;
 
public class RectangularPasture {

...

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[] xs = new int[n];
        int[] ys = new int[n];
        Integer[] cows = new Integer[n];
        for (int j = 0; j < n; j++) {
            xs[j] = in.nextInt();
            ys[j] = in.nextInt();
            cows[j] = j;
        }
        Arrays.sort(cows, Comparator.comparingInt(j -> xs[j]));
        for (int x = 1; x <= n; x++) {
            xs[cows[x - 1]] = x;
        }
        Arrays.sort(cows, Comparator.comparingInt(j -> ys[j]));
        for (int y = 1; y <= n; y++) {
            ys[cows[y - 1]] = y;
        }

...

```

The solution uses a Lambda function as the custom comparator, which our guide didn't discuss, but it should be apparent which coordinate (`x` or `y`) that the comparator is sorting by.

</JavaSection>

<PySection>

<IncompleteSection>

Sorry, there is no python solution. If you have a python solution to the problem (that uses coordinate compression), we would greatly appreciate if you contributed it!

</IncompleteSection>

> </PySection>

Reviewer: i think if u delete this the incomplete section will be there
